src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentReadStream.cs

```diff
@@ -38,5 +38,41 @@ public sealed override Task<int> ReadAsync(byte[] buffer, int offset, int count,
 
         public sealed override void CopyTo(Stream destination, int bufferSize) =>
             CopyToAsync(destination, bufferSize, CancellationToken.None).GetAwaiter().GetResult();
+
+        public virtual Task DrainAsync(int maxDrainBytes) => Task.CompletedTask;
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                if (_connection != null)
+                {
+                    // Start the asynchronous drain.
+                    // It may complete synchronously, in which case the connection will be put back in the pool synchronously.
+                    // Skip the call to base.Dispose -- it will be deferred until DrainOnDispose finishes.
+                    DrainOnDispose();
+                    return;
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+        // Maximum request drain size, 16K.
+        private const int s_maxDrainBytes = 16 * 1024;
```

Comment:
Nit: s_maxDrainBytes => MaxDrainBytes.  Unless you plan to expose a public API to configure this? (WinHttpHandler has MaxResponseDrainSize.)

Reply:
Also, WinHttpHandler defaults to 1MB, as that's the default for WINHTTP_OPTION_MAX_RESPONSE_DRAIN_SIZE.  Without concrete data that would help us pick a better alternative, maybe go with that?

Reply:
Native WinHTTP defaults to 1MB.  However, we went with a different setting in WinHttpHandler, 64K which was based on .NET Framework.

However, we ended up with a bug https://github.com/dotnet/corefx/issues/18000.  This was caused by the 3xx redirect response having a large response body (greater than 64K) and WinHTTP couldn't drain that much (since we set the default to 64k).  But then WinHTTP has a separate bug and didn't create new connection.  Discussions with WinHTTP team said that we should have just left it set at 1MB since that works well in most real-world scenarios.

WinHttpHandler is still using 64KB but probably should be changed to 1MB.

Reply:
I wasn't planning on exposing it publicly, at least for now.  I filed #27329 to track.

Reply:
I'm fine with 1MB.  Will change.

@davidsh, in practice, when chunked encoding is used, I'm not seeing WinHttpHandler drain up to 1MB -- in some cases it doesn't even seem to be draining more than a few kilobytes.  Testing this is tricky so I may be doing something wrong in the tests.  Do you know what's going on here?

Reply:
WinHttpHandler doesn't do any draining.  Connection management is handled in the lower native WinHTTP layer itself.  I'm not sure why WinHTTP might behave that way w.r.t. reading chunked encoding responses.  But you can ask the WinHTTP team about this.